### ISSUE TITLE
Autosaving Registration Form

### ISSUE THREAD

**Issue by wiput1999** [ycccamp/ycc-regis#1]
The autosaving objective is when user is typing content and sometimes user accident close tab or browser then they need to write all of it again.

**Comment by SaltyAom** [comment]
Since I had problem with source-code's complexity and structure plus I have never used `Formik` before, instead I'll leave some advice instead since I wouldn't wanted to break the source code.

Call `enablePersistence()` on Firestore first (you had already did) and once necessary information if fetched, call `firestore.disableNetwork()` once you do that, so might call `set()` and `update()` every *single* time form is updated. This'll save every data as cache to Firestore's IndexedDB but won't update to Firebase, this would prevent write count. Once form is ready to submit. Just call firebase.enableNetwork()
```javascript
let instance = firebase.firestore()
instance.disableNetwork()
...

instance...set({ ... })
...

// Without set/update, it would just silently submit data to Firestore
instance.enableNetwork()

// Prevent silently data from being submit by overwrite write queue
instance...set(finalData)
```

In case I'll have to access and manipulate the code myself, I'll need some help.
1) Since I have never used `formik` before and have some problem with complexity of source-code's structure plus I can't get the form to be displayed. Which if I had get down and manipulate the code myself, I would need some advice. I found components like `FormDate` `FormInput` which I'm not quite sure how should I manipulate onChange event since it received props `formik.handleChange` which I'm not quite sure if I should changed it. (Again, I couldn't test any code yet.)
  
2) I can't access to registration form since my Facebook account hasn't been added for testing. Maybe someone could help me about that. (I have try modifying some source-code turns out there's a lot of wire preventing me from displaying the form)
<img width="1440" alt="Screen Shot 2563-01-20 at 15 14 13" src="https://user-images.githubusercontent.com/35027979/72709422-85da5400-3b97-11ea-95f7-5f08ff04c9a3.png">
  
3) Since I has inspect some source-code, this is an opinion about optimization please don't take it too serious thus we could do it later.
  
3.1) Shouldn't we use `async function` in `useEffect()` to prevent callback hell?
```javascript
useEffect(() => {
  // Handle code asychrnously
  (async (){
      let user = await instance
        .firestore()
        .collection('registration')
        .doc(user.uid)
        .collection('forms')
        .doc('personal')
        .get()

    setForm(user.data())
  })()
}, [])
```
3.2) `instance.firestore()` and `instance.storage()` should only call once.  
3.3) I think we should `lazy load` the firebase module, since it's very heavy (100kB+). There's a documentation about how to do that on Next.js website.
  
###### Note: I forked from current master.

**Comment by rayriffy** [comment]
For auth, you can temporarily fix by changing provider to `GoogleAuthProvider`. This issue occurs because Facebook Login require HTTPS to authenticate via OAuth which development environment is not.
https://github.com/ycccamp/ycc-regis/blob/master/src/app/components/auth.tsx#L24

For a record, you can refactor entire thing as you wish since I am still busy on [ycc.in.th](https://ycc.in.th)

**Comment by wiput1999** [comment]
@SaltyAom @rayriffy https://gist.github.com/jaredpalmer/56e10cabe839747b84b81410839829be I see this example

**Comment by SaltyAom** [comment]
Created a pull request for auto-saving #2
I'd recommended adding a debounce function.
Also be sure to check if there's something wrong but I tested it on my local development environment.

**Comment by SaltyAom** [comment]
It use basic localStorage under the hood.

**Comment by SaltyAom** [comment]
So, I adding [this feature](https://github.com/ycccamp/ycc-regis/pull/2#issuecomment-576503924) which I think I need to access Firestore to check the database structure.
Otherwise, you could just add the timestamp at root level and compare it to localstorage.
<img alt="Screen Shot 2563-01-26 at 10 33 18" src="https://user-images.githubusercontent.com/35027979/73130246-5019ed00-4027-11ea-8e0e-7764981216ca.png">

**Comment by rayriffy** [comment]
I added you to a project now with email `ao********3@gmail.com`, please check your box

**Comment by SaltyAom** [comment]
Due to #2 is merged into master, I would like to close the issue. Thanks for your patience!
